### ISSUE TITLE
Cross Modality Between CT & MRI

### ISSUE THREAD

**Issue by deltarfd** [MIC-DKFZ/nnUNet#383]
Hi @FabianIsensee ,

I am trying to segment the image using dataset from [CHAOS - Combined (CT-MR) Healthy Abdominal Organ Segmentation](https://chaos.grand-challenge.org/). Those that have been available and have been tried on nnUNet are only limited to MRI modalities, but is it possible to segment with CT and MRI modalities using nnUNet? Thank you.

Best,
Delta

**Comment by FabianIsensee** [comment]
Hi,
I have not tried that but it should work. The thing is that if you throw them all into one dataset like I did you are effectively treating all the images as one modality (and I would not recommend naming this modality CT in the dataset.json!). This means that the CT images will be normalized with image-level mean and standard deviation. I think this work well enough, though.
Best,
Fabian

**Comment by bkonk** [comment]
Would you recommend doing any pre-preprocessing on the CT data? For example, window-leveling the pixel values and then feeding this in together with raw unaltered MR data for the multi-modal training.

**Comment by FabianIsensee** [comment]
I would try both

**Comment by deltarfd** [comment]
Hi @FabianIsensee ,
When I did this task, I found 3 patients in the MR dataset (Patients 21,22, and 33) had different geometries (does not match the geometry of the label file). How do you solve the problem? Whether to remove the verify dataset integrity flag ?
Thank you for your reply!
Best,
Delta

**Comment by FabianIsensee** [comment]
Hi,
honestly I don't know whether I even noticed. Which cases are those?
Best,
Fabain

**Comment by deltarfd** [comment]
For all cases in the patient number 21, 22, and 33 (T2, T1 Inphase, and  T1 Outphase).

checking case T1_out_21
the direction does not match between the images
checking case T1_in_21
the direction does not match between the images
checking case T1_out_21
the direction does not match between the images
checking case T1_out_22
the direction does not match between the images
checking case T1_in_22
the direction does not match between the images
checking case T1_out_33
the direction does not match between the images
checking case T1_in_33
the direction does not match between the images
checking case T2_21
the direction does not match between the images
checking case T2_22
the direction does not match between the images
checking case T2_33
the direction does not match between the images

**Comment by FabianIsensee** [comment]
I will have a look, but not today. It's too busy at the moment. Sometimes the numeric values of the affine are not 100% identical. This will print this warning even though it is not a problem at all in practice.
You can verify that yourself by looking at the .npz files generated by cropping. If seg and data overlap in there then it's all good.

**Comment by deltarfd** [comment]
Thank you! Waiting for any further information from you.

**Comment by MatinHz** [comment]
Hi @FabianIsensee,

ADC in prostate mpMRI dataset is also a quantitative image like CT. 
Would it be okay if I change ADC in dataset.json to CT in order to use global normalization and use nonCT normalization for another image at the same time? (json_dict['modality'] = { "0": "MRI", "1": "CT"})

Best,
Matin

**Comment by FabianIsensee** [comment]
Hi @MatinHz , this should be a separate issue. But while we are at it: Yes you can set it to CT. This only affects the intensity normalization during preprocessing. If I were you I would try both just to see whether this actually works well

**Comment by MatinHz** [comment]
Thanks @FabianIsensee!

**Comment by FabianIsensee** [comment]
@deltarfd I checked the case and they all look good to me. This is what I used to check the preprocessed images:

```python
from batchviewer import view_batch
import numpy as np
a = np.load('T2_33.npy')
view_batch(a, np.clip(a[-1], a_min=0, a_max=999) + a[0])
```

(https://github.com/FabianIsensee/BatchViewer)
Best,
Fabian

**Comment by FabianIsensee** [comment]
`the direction does not match between the images` is intended as a warning and should prompt users to have a closer look. It does not mean that there is definitely a problems

**Comment by deltarfd** [comment]
@FabianIsensee Thank you so much for helping to check the case. I am still working on the experiment.

**Comment by FabianIsensee** [comment]
any updates?

**Comment by deltarfd** [comment]
Hello @FabianIsensee,
I've tried running training with 2d and 3d_fullres using foldable 'all' with 200 epochs. I have tried submit with the results:
2d: https://chaos.grand-challenge.org/evaluation/6b9592f3-dcc2-46d6-babd-ced7226c96f9/
3d: https://chaos.grand-challenge.org/evaluation/7f6678a2-87f6-4537-a26b-f2ce9d22c6c5/

In these results, there is a blank value on the CT Dice Score. Is there something wrong with what I did? I used the same convert intensity for Task 1 and 4 as your program (Task 3 and 5). Are there any suggestions for me to improve? Thank you

**Comment by FabianIsensee** [comment]
Hi, difficult to say. Ideally you would just visually inspect the images to see if there is something obviously wrong.
Best,
Fabian

**Comment by deltarfd** [comment]
Hello, I want to ask again. I think the training error is due to the wrong order of the images. SITK creates a reverse sequence of CT images. After I fixed, why is my submit result evaluated with "_Failed Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample._" Any suggestions with this error?
Thank you.

**Comment by FabianIsensee** [comment]
Run `nnUNet_plot_task_pngs` to confirm that images and segmentations match.

If you get errors: Make sure your modalities + their segmentation are aligned. nnU-Net ignores all geometry information (except spacing). So when you extract the numpy arrays from image and seg then they need to have the same shape and they need to overlap correctly

**Comment by deltarfd** [comment]
This is the result of the command `nnUNet_plot_task_pngs -t 101 -o Plot_PNG --use_raw` that I ran. And the result is like [here](https://drive.google.com/drive/folders/1N9Buh5fgB38cRmlsCGhXti7ujVHLOmdL?usp=sharing), in the `Plot_PNG` folder. In that folder I also paste the segmentation results in the `CHAOS_submission_template_new` file, and the code I modified from your code (`Task101_Chaos_Challenge.py` as Dataset Conversion and `convert_result.py` as submit format conversion). Maybe you can easily find the mistake I did and tell me the right one. Thank you

**Comment by FabianIsensee** [comment]
I do not have the capacities to run a code review on your script. Your png files look good, so that's not it. Does the error occur when you try to submit?

**Comment by deltarfd** [comment]
Yes, the last time I submitted an error occurred as I said above. And I don't know where the error lies. If you find an error, maybe you can let me know because I really hope someone can help.

**Comment by deltarfd** [comment]
Hello, I've found the error. Sorry, I wasn't careful enough that I changed the intensity at the time of `convert_predicted`. The thing I will ask further is whether I can do ensembles without `plans.pkl`, because here I do as you do (Variant 2). I've done train with 2D and 3D with 5 fold, and I've ensembled. But I failed when I converted the ensemble result, because I needed `plans.pkl `
```
shutil.copy(join(folder_with_predictions, "plans.pkl"), join(folder_for_ensembing0, "plans.pkl"))
shutil.copy(join(folder_with_predictions, "plans.pkl"), join(folder_for_ensembing1, "plans.pkl"))
```
How do I do `convert_predicted `without it?
Thank you
